[WIP] OSDOCS:2475 Adds new assembly for the dynamic plugin TP (#41260)
openshift/openshift-docs: modules/running-your-dynamic-plug-in.adoc

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="running-your-dynamic-plugin"]
= Running your dynamic plug-in
You can run the plugin using a local development environment.

== Run your plug-in locally

.Procedure

> . Run `bridge` locally and instruct it to proxy requests directly to your local plug-in asset server (web server hosting the plugin's generated assets): `./bin/bridge -plugins console-demo-plugin=http://localhost:9001/`

Reviewer: Hey, @opayne1. I'm not sure if you're still working on this, but running bridge this way is no longer necessary. We should follow the steps laid out in the local development section of the template README here:

https://github.com/spadgett/console-plugin-template#option-1-local

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="running-your-dynamic-plugin_{context}"]
= Running your dynamic plug-in

You can run the plugin using a local development environment. The OpenShift console will run in a container connected to the cluster you have logged into.

.Prerequisites
* You must have the OpenShift CLI (`oc`) installed.
* You must have an OpenShift cluster running.

> * You must have link:https://www.docker.com/[Docker] (for MAC) or link:https://podman.io/[Podman] (for RHEL) installed.

Reviewer: I don't know if we have these recommendations in other places, but it's possible to use podman for mac.

.Procedure

. In one terminal window, run:
+
[source,terminal]
----
$ yarn install
----

. Run:
+
[source,terminal]
----
$ yarn run start
----

> . In the OpenShift CLI, run:

Reviewer: ```suggestion
. In a second terminal window, run:
```

We probably want to restart the numbering here since these commands are run in a different window.

> b. Run the following commands in the second terminal window.

Reviewer 2: ```suggestion
b. Run the following commands in the second terminal window:
```

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="running-your-dynamic-plugin_{context}"]
= Running your dynamic plug-in

> You can run the plugin using a local development environment. The OpenShift console will run in a container connected to the cluster you have logged into.

Reviewer 2: ```suggestion
You can run the plug-in using a local development environment. The OpenShift console runs in a container connected to the cluster you have logged into.
```

.Prerequisites
* You must have the OpenShift CLI (`oc`) installed.
* You must have an OpenShift cluster running.
* You must have link:https://www.docker.com/[Docker] or link:https://podman.io/[Podman] installed.

.Procedure

> In one terminal window:

Reviewer 3: Shall we say 
```
In the local cloned directory,in one terminal window
```

Reviewer 2: What do you think of this:

1. Open two terminal windows in the local directory of your cloned repository.
2. Run the following commands in the first terminal:
3. Run the following commands in the second terminal:

.Verification
- Visit localhost...

Author: I like that

. Run the following commands:
+
[source,terminal]
----
$ yarn install
----
+
[source,terminal]
----
$ yarn run start
----

In a second terminal window:

[start=1]
. Run the following commands:
+
[source,terminal]
----
$ oc login
----
+
[source,terminal]
----
$ yarn run start-console
----

> . Visit link:http://localhost:9000/example[local host] to view the running plug-in.

Reviewer 3: shall we make `http://localhost:9000/example` plain link here? since some user may trying on a different machine and he can only knows the actual link by clicking `local host`, if we write `http://localhost:9000/example` here it would be quite straightforward

Reviewer 3: Also I followed the exact same steps in this doc and I got following `404 Not Found` error when visiting `http://localhost:9000/example` , also I see some errors like
```
Error while loading plugin from /api/plugins/console-plugin-template/ a: GET request for "console-plugin-template" plugin failed: Get "http://host.containers.internal:9001/plugin-manifest.json": dial tcp 10.0.2.2:9001: connect: network is unreachable
```

Reviewer: @yapei What podman version are you using? I suspect you might need to upgrade podman.

Reviewer: You need at least 3.2.0: https://github.com/containers/podman/releases/tag/v3.2.0

> Podman now adds an entry to /etc/hosts, host.containers.internal, pointing to the current gateway (which, for root containers, is usually a bridge interface on the host system)

Author: @spadgett should I add 3.2.0 by Podman for clarity?

Reviewer: @opayne1 I think so. I added it to the plugin template repo README as well.

Reviewer 3: @spadgett  I'm running on Fedora and Docker is expected to be used, right? 
My docker version on fedora is:
```
$ docker version
Client:
 Version:           20.10.10
 API version:       1.41
 Go version:        go1.16.8
 Git commit:        485636f
 Built:             Sat Nov 13 14:50:36 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

```

My Podman version on fedora is:
```
$ podman version
Version:      3.4.2
API Version:  3.4.2
Go Version:   go1.16.8
Built:        Fri Nov 12 20:25:37 2021
OS/Arch:      linux/amd64
```

// Module included in the following assemblies:
//
// * web_console/dynamic-plug-ins.adoc

:_content-type: PROCEDURE
[id="running-your-dynamic-plugin"]
= Running your dynamic plug-in
You can run the plugin using a local development environment. The OpenShift console will run in a container connected to the cluster you have logged into.

[id="run-plug-in-locally"]
== Run your plug-in locally

.Prerequisites

> * You must have the OpenShift CLI link:https://mirror.openshift.com/pub/[(`oc`)] installed.

Reviewer: I'm not sure the best link for downloading `oc`, but I don't know that it's mirror.openshift.com. I think it's reasonable to assume they already have it and not have a link.

* You must have an OpenShift cluster running.

> * You must have link:https://www.docker.com/[Docker] or link:https://podman.io/[Podman] installed.

Reviewer: We should also check here about how we want to handle Docker vs podman.

Author: I added mac/RHEL but I am not sure if that is what you are looking for.

Author: [Here]( https://docs.openshift.com/container-platform/4.9/openshift_images/index.html#images-about_overview-of-images) is how we have docker/podman in other places of the docs. It looks like users can choose either or. Is that the case for this?

Reviewer: Yes, it's possible for developers to use either, and the commands are otherwise the same.

Reviewer 3: Shall we say Docker/Podman should be running? @opayne1 
```
* You must have link:https://www.docker.com/[Docker] or link:https://podman.io/[Podman] installed and running
```

.Procedure

. In one terminal window, run:
+

[source,terminal]
----
yarn install
----

. Run:
+

[source,terminal]
----
yarn run start
----

. In the OpenShift CLI, run:
+

[source,terminal]
----
oc login
----

. Run:
+

[source,terminal]
----
yarn run start-console
----

. Visit link:http://localhost:9000/example[local host] to view the running plug-in.

[id="run-plug-in-docker-and-VSCode-remote-container"]

> == Run your plug-in with Docker and VSCode Remote Container

Reviewer: I would probably leave this out for simplicity and only cover the first way to run the plugin.
